master/all/man/man3/KeyAction.3
.TH "KeyAction" 3 "Sat Jan 20 2018" "Version v0.2.8 at branch master" "ckb-next" \" -*- nroff -*-
.ad l
.nh
.SH NAME
KeyAction \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <src/ckb/keyaction\&.h>\fP
.PP
Inherits \fBQObject\fP\&.
.SS "Public Types"

.in +1c
.ti -1c
.RI "enum \fBType\fP { \fBUNBOUND\fP, \fBNORMAL\fP, \fBSPECIAL\fP }"
.br
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBKeyAction\fP (const QString &action, \fBQObject\fP *parent=0)"
.br
.ti -1c
.RI "QString \fBvalue\fP () const "
.br
.ti -1c
.RI "\fBoperator QString\fP () const "
.br
.ti -1c
.RI "\fBKeyAction\fP (\fBQObject\fP *parent=0)"
.br
.ti -1c
.RI "QString \fBfriendlyName\fP (const \fBKeyMap\fP &map) const "
.br
.ti -1c
.RI "QString \fBdriverName\fP () const "
.br
.ti -1c
.RI "QString \fBmacroFullLine\fP () const "
.br
.RI "\fImacroFullLine If a macro command and a macro definition exists for the given key, returns the complete string except the leading '$' (the $ may confuse some caller)\&. \fP"
.ti -1c
.RI "bool \fBisValidMacro\fP () const "
.br
.RI "\fIisValidMacro checks whether a keyAction contains a valid macro\&. This is done easily: If the macro action starts with $macro: and has five elements, delimited by ':', we may assume, that is a structural correct macro action\&. If it has 4 entries only, it is an older definition and ok also\&. \fP"
.ti -1c
.RI "QStringList \fBmacroLine\fP () const "
.br
.RI "\fImacroLine returns all interresting content for a macro definition\&. \fP"
.ti -1c
.RI "QString \fBmacroContent\fP () const "
.br
.RI "\fImacroContent returns the macro key definition only (the second part of the macro action)\&. \fP"
.ti -1c
.RI "QString \fBmacroTiming\fP () const "
.br
.RI "\fImacroTiming returns the macro key definition with original timing infos (the fifth and up to now last part of the macro action)\&. If the implementation does not know anything about delays and has no 5th part, return first part\&. \fP"
.ti -1c
.RI "void \fBmacroDisplay\fP ()"
.br
.RI "\fIDebug output for invalid macro Definitions\&. \fP"
.ti -1c
.RI "\fBType\fP \fBtype\fP () const "
.br
.ti -1c
.RI "bool \fBisUnbound\fP () const "
.br
.ti -1c
.RI "bool \fBisNormal\fP () const "
.br
.ti -1c
.RI "bool \fBisSpecial\fP () const "
.br
.ti -1c
.RI "bool \fBisMedia\fP () const "
.br
.ti -1c
.RI "bool \fBisProgram\fP () const "
.br
.ti -1c
.RI "bool \fBisAnim\fP () const "
.br
.ti -1c
.RI "bool \fBisMacro\fP () const "
.br
.ti -1c
.RI "bool \fBisDPI\fP () const "
.br
.ti -1c
.RI "bool \fBisMouse\fP () const "
.br
.ti -1c
.RI "QString \fBspecialInfo\fP (int &\fBparameter\fP) const "
.br
.ti -1c
.RI "int \fBprogramInfo\fP (QString &onPress, QString &onRelease) const "
.br
.ti -1c
.RI "int \fBdpiInfo\fP (QPoint &custom) const "
.br
.ti -1c
.RI "QUuid \fBanimInfo\fP (bool &onlyOnce, bool &stopOnRelease) const "
.br
.ti -1c
.RI "void \fBkeyEvent\fP (\fBKbBind\fP *bind, bool down)"
.br
.ti -1c
.RI "void \fBkeyRelease\fP (\fBKbBind\fP *bind)"
.br
.ti -1c
.RI "void \fBadjustDisplay\fP ()"
.br
.ti -1c
.RI "\fB~KeyAction\fP ()"
.br
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static QString \fBnoAction\fP ()"
.br
.ti -1c
.RI "static QString \fBdefaultAction\fP (const QString &\fBkey\fP, \fBKeyMap::Model\fP model)"
.br
.ti -1c
.RI "static QString \fBmodeAction\fP (int mode)"
.br
.ti -1c
.RI "static QString \fBdpiAction\fP (int level, int customX=0, int customY=0)"
.br
.ti -1c
.RI "static QString \fBlightAction\fP (int \fBtype\fP=\fBLIGHT_UP_WRAP\fP)"
.br
.ti -1c
.RI "static QString \fBlockAction\fP (int \fBtype\fP=\fBLOCK_TOGGLE\fP)"
.br
.ti -1c
.RI "static QString \fBprogramAction\fP (const QString &onPress, const QString &onRelease, int stop)"
.br
.ti -1c
.RI "static QString \fBanimAction\fP (const QUuid &guid, bool onlyOnce, bool stopOnRelease)"
.br
.ti -1c
.RI "static QString \fBmacroAction\fP (QString macroDef)"
.br
.RI "\fIwell documented in cpp file \fP"
.in -1c
.SS "Static Public Attributes"

.in +1c
.ti -1c
.RI "static const int \fBMODE_PREV\fP = -2"
.br
.ti -1c
.RI "static const int \fBMODE_NEXT\fP = -1"
.br
.ti -1c
.RI "static const int \fBMODE_PREV_WRAP\fP = -4"
.br
.ti -1c
.RI "static const int \fBMODE_NEXT_WRAP\fP = -3"
.br
.ti -1c
.RI "static const int \fBDPI_CYCLE_UP\fP = -4"
.br
.ti -1c
.RI "static const int \fBDPI_CYCLE_DOWN\fP = -3"
.br
.ti -1c
.RI "static const int \fBDPI_UP\fP = -2"
.br
.ti -1c
.RI "static const int \fBDPI_DOWN\fP = -1"
.br
.ti -1c
.RI "static const int \fBDPI_SNIPER\fP = 0"
.br
.ti -1c
.RI "static const int \fBDPI_CUSTOM\fP = 6"
.br
.ti -1c
.RI "static const int \fBLIGHT_UP\fP = 0"
.br
.ti -1c
.RI "static const int \fBLIGHT_DOWN\fP = 1"
.br
.ti -1c
.RI "static const int \fBLIGHT_UP_WRAP\fP = 2"
.br
.ti -1c
.RI "static const int \fBLIGHT_DOWN_WRAP\fP = 3"
.br
.ti -1c
.RI "static const int \fBLOCK_TOGGLE\fP = 0"
.br
.ti -1c
.RI "static const int \fBLOCK_ON\fP = 1"
.br
.ti -1c
.RI "static const int \fBLOCK_OFF\fP = 2"
.br
.ti -1c
.RI "static const int \fBPROGRAM_PR_MULTI\fP = 0x04"
.br
.ti -1c
.RI "static const int \fBPROGRAM_PR_INDEF\fP = 0x00"
.br
.ti -1c
.RI "static const int \fBPROGRAM_PR_KRSTOP\fP = 0x01"
.br
.ti -1c
.RI "static const int \fBPROGRAM_PR_KPSTOP\fP = 0x02"
.br
.ti -1c
.RI "static const int \fBPROGRAM_RE_MULTI\fP = 0x40"
.br
.ti -1c
.RI "static const int \fBPROGRAM_RE_INDEF\fP = 0x00"
.br
.ti -1c
.RI "static const int \fBPROGRAM_RE_KPSTOP\fP = 0x20"
.br
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "void \fBoperator=\fP (const \fBKeyAction\fP &rhs)"
.br
.RI "\fIccMSC: Don't copy key actions (the old one needs to be deleted first) frickler24: statement left as described, but copying is done in \fBKbBind\fP copy constructor \fP"
.ti -1c
.RI "\fBKeyAction\fP (const \fBKeyAction\fP &rhs)"
.br
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "QString \fB_value\fP"
.br
.ti -1c
.RI "QProcess * \fBpreProgram\fP"
.br
.ti -1c
.RI "QProcess * \fBrelProgram\fP"
.br
.ti -1c
.RI "quint64 \fBsniperValue\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
Definition at line 13 of file keyaction\&.h\&.
.SH "Member Enumeration Documentation"
.PP 
.SS "enum \fBKeyAction::Type\fP"

.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fIUNBOUND \fP\fP
.TP
\fB\fINORMAL \fP\fP
.TP
\fB\fISPECIAL \fP\fP
.PP
Definition at line 146 of file keyaction\&.h\&.
.PP
.nf
146               {
147         UNBOUND,
148         NORMAL,
149         SPECIAL,
150     };
.fi
.SH "Constructor & Destructor Documentation"
.PP 
.SS "KeyAction::KeyAction (const QString &action, \fBQObject\fP *parent = \fC0\fP)"

.PP
Definition at line 22 of file keyaction\&.cpp\&.
.PP
.nf
23     : QObject(parent), _value(action), preProgram(0), relProgram(0), sniperValue(0)
24 {
25 }
.fi
.SS "KeyAction::KeyAction (\fBQObject\fP *parent = \fC0\fP)\fC [explicit]\fP"

.PP
Definition at line 27 of file keyaction\&.cpp\&.
.PP
.nf
28     : QObject(parent), _value(""), preProgram(0), relProgram(0), sniperValue(0)
29 {
30 }
.fi
.SS "KeyAction::~KeyAction ()"

.PP
Definition at line 32 of file keyaction\&.cpp\&.
.PP
References preProgram, and relProgram\&.
.PP
.nf
32                      {
33     // Clean up processes
34     if(preProgram){
35         preProgram->kill();
36         delete preProgram;
37     }
38     if(relProgram){
39         relProgram->kill();
40         delete relProgram;
41     }
42 }
.fi
.SS "KeyAction::KeyAction (const \fBKeyAction\fP &rhs)\fC [inline]\fP, \fC [private]\fP"

.PP
Definition at line 187 of file keyaction\&.h\&.
.PP
.nf
187 : QObject() {}
.fi
.SH "Member Function Documentation"
.PP 
.SS "void KeyAction::adjustDisplay ()"

.PP
Definition at line 464 of file keyaction\&.cpp\&.
.PP
Referenced by keyEvent()\&.
.PP
.nf
464                              {
465 #ifdef USE_LIBX11
466     // Try to get the current display from the X server
467     char* display_name = XDisplayName(NULL);
468     if(!display_name)
469         return;
470     Display* display = XOpenDisplay(display_name);
471     if(!display)
472         return;
473     char* display_string = DisplayString(display);
474     if(!display_string || strlen(display_string) == 0){
475         XCloseDisplay(display);
476         return;
477     }
478     size_t envstr_size = strlen(display_string) + 4;
479     char* envstr = new char[envstr_size];
480     strncpy(envstr, display_string, envstr_size);
481     envstr[envstr_size - 1] = 0;
482 
483     Window root_window = XRootWindow(display, DefaultScreen(display));
484     Window root_window_ret, child_window_ret, window;
485     XWindowAttributes attr;
486     int root_x, root_y, win_x, win_y;
487     unsigned int mask_ret;
488 
489     // Find the screen which currently has the mouse
490     XQueryPointer(display, root_window, &root_window_ret, &child_window_ret, &root_x, &root_y, &win_x, &win_y, &mask_ret);
491     if(child_window_ret == (Window)NULL)
492         window = root_window_ret;
493     else
494         window = child_window_ret;
495     XGetWindowAttributes(display, window,  &attr);
496 
497     char* ptr = strchr(envstr, ':');
498     if(ptr){
499         ptr = strchr(ptr, '\&.');
500         if(ptr)
501             *ptr = '\0';
502         char buf[16];
503         snprintf(buf, sizeof(buf), "\&.%i", XScreenNumberOfScreen(attr\&.screen));
504         strncat(envstr, buf, envstr_size - 1 - strlen(envstr));
505 
506         // Update environment variable
507         setenv("DISPLAY", envstr, 1);
508     }
509 
510     delete[] envstr;
511     XCloseDisplay(display);
512 #endif
513 }
.fi
.SS "QString KeyAction::animAction (const QUuid &guid, boolonlyOnce, boolstopOnRelease)\fC [static]\fP"

.PP
Definition at line 184 of file keyaction\&.cpp\&.
.PP
References ANIM_KRSTOP, and ANIM_ONCE\&.
.PP
Referenced by RebindWidget::applyChanges()\&.
.PP
.nf
184                                                                                  {
185     int flags = (onlyOnce ? ANIM_ONCE : 0) | (stopOnRelease ? ANIM_KRSTOP : 0);
186     return "$anim:" + guid\&.toString() + QString("+%1")\&.arg(flags);
187 }
.fi
.SS "QUuid KeyAction::animInfo (bool &onlyOnce, bool &stopOnRelease) const"

.PP
Definition at line 225 of file keyaction\&.cpp\&.
.PP
References _value, ANIM_KRSTOP, ANIM_ONCE, and isAnim()\&.
.PP
Referenced by keyEvent(), and RebindWidget::setSelection()\&.
.PP
.nf
225                                                                    {
226     if(!isAnim())
227         return QUuid();
228     QString param = _value\&.mid(6);
229     QStringList split = param\&.split("+");
230     if(split\&.length() < 2)
231         return QUuid();
232     QUuid id = split[0];
233     int flags = split[1]\&.toInt();
234     onlyOnce = !!(flags & ANIM_ONCE);
235     stopOnRelease = !!(flags & ANIM_KRSTOP);
236     return id;
237 }
.fi
.SS "QString KeyAction::defaultAction (const QString &key, \fBKeyMap::Model\fPmodel)\fC [static]\fP"

.PP
Definition at line 44 of file keyaction\&.cpp\&.
.PP
References KeyMap::GLAIVE, and KeyMap::HARPOON\&.
.PP
Referenced by KbBind::bindAction(), KbBind::defaultAction(), and KbBind::save()\&.
.PP
.nf
44                                                                      {
45     // G1-G18 are unbound by default
46     if(key\&.length() >= 2 && key[0] == 'g'
47         && ((key\&.length() == 2 && key[1] >= '0' && key[1] <= '9')
48         || (key\&.length() == 3 && key[1] == '1' && key[2] >= '0' && key[2] <= '8')))
49         return "";
50     // So are thumbgrid buttons
51     if(key\&.startsWith("thumb"))
52         return "";
53     // TODO: default action for MR
54     if(key == "mr")
55         return "";
56     // M1-M3 switch modes
57     if(key == "m1")
58         return "$mode:0";
59     if(key == "m2")
60         return "$mode:1";
61     if(key == "m3")
62         return "$mode:2";
63     // Brightness and Win Lock are their own functions
64     if(key == "light")
65         return "$light:2";
66     if(key == "lock")
67         return "$lock:0";
68     // DPI buttons
69     if(key == "dpiup"){
70         if(model == KeyMap::HARPOON ||
71        model == KeyMap::GLAIVE){
72         return "$dpi:-4";
73         }
74     return "$dpi:-2";
75     }
76     if(key == "dpidn")
77         return "$dpi:-1";
78     if(key == "sniper")
79         return "$dpi:0";
80     // Everything else is a standard keypress
81     return key;
82 }
.fi
.SS "QString KeyAction::dpiAction (intlevel, intcustomX = \fC0\fP, intcustomY = \fC0\fP)\fC [static]\fP"

.PP
Definition at line 162 of file keyaction\&.cpp\&.
.PP
References DPI_CUSTOM\&.
.PP
Referenced by RebindWidget::applyChanges()\&.
.PP
.nf
162                                                                {
163     QString action = tr("$dpi:%1")\&.arg(level);
164     if(level == DPI_CUSTOM)
165         action += tr("+%1+%2")\&.arg(customX)\&.arg(customY);
166     return action;
167 }
.fi
.SS "int KeyAction::dpiInfo (QPoint &custom) const"

.PP
Definition at line 211 of file keyaction\&.cpp\&.
.PP
References _value, DPI_CUSTOM, and isDPI()\&.
.PP
Referenced by friendlyName(), keyEvent(), and RebindWidget::setSelection()\&.
.PP
.nf
211                                            {
212     if(!isDPI())
213         return 0;
214     QString param = _value\&.mid(5);
215     QStringList lxy = param\&.split("+");
216     int level = lxy[0]\&.toInt();
217     if(level == DPI_CUSTOM){
218         if(lxy\&.length() != 3)
219             return 0;
220         custom = QPoint(lxy[1]\&.toInt(), lxy[2]\&.toInt());
221     }
222     return level;
223 }
.fi
.SS "QString KeyAction::driverName () const"

.PP
Definition at line 239 of file keyaction\&.cpp\&.
.PP
References _value, and isSpecial()\&.
.PP
Referenced by KbBind::update()\&.
.PP
.nf
239                                     {
240     if(isSpecial())
241         return "";
242     return _value;
243 }
.fi
.SS "QString KeyAction::friendlyName (const \fBKeyMap\fP &map) const"

.PP
Definition at line 84 of file keyaction\&.cpp\&.
.PP
References _value, DPI_CUSTOM, DPI_CYCLE_DOWN, DPI_CYCLE_UP, DPI_DOWN, DPI_SNIPER, DPI_UP, dpiInfo(), KeyMap::friendlyName(), KeyMap::layout(), LIGHT_DOWN, LIGHT_DOWN_WRAP, LIGHT_UP, LIGHT_UP_WRAP, LOCK_OFF, LOCK_ON, LOCK_TOGGLE, MODE_NEXT, MODE_NEXT_WRAP, MODE_PREV, and MODE_PREV_WRAP\&.
.PP
.nf
84                                                        {
85     if(_value\&.isEmpty())
86         return "Unbound";
87     QStringList parts = _value\&.split(":");
88     QString prefix = parts[0];
89     if(parts\&.length() < 2){
90         KeyMap::Layout layout = map\&.layout();
91         QString name = KeyMap::friendlyName(_value, layout);
92         if(name\&.isEmpty())
93             return "(Unknown)";
94         return name;
95     }
96     int suffix = parts[1]\&.toInt();
97     if(prefix == "$mode"){
98         switch(suffix){
99         case MODE_PREV:
100         case MODE_PREV_WRAP:
101             return "Switch to previous mode";
102         case MODE_NEXT:
103         case MODE_NEXT_WRAP:
104             return "Switch to next mode";
105         default:
106             return tr("Switch to mode %1")\&.arg(suffix + 1);
107         }
108     } else if(prefix == "$dpi"){
109         // Split off custom parameters (if any)
110         int level = parts[1]\&.split("+")[0]\&.toInt();
111         switch(level){
112         case DPI_CYCLE_UP:
113             return "DPI cycle up";
114         case DPI_CYCLE_DOWN:
115             return "DPI cycle down";
116         case DPI_UP:
117             return "DPI up";
118         case DPI_DOWN:
119             return "DPI down";
120         case DPI_SNIPER:
121             return "Sniper";
122         case DPI_CUSTOM:{
123             QPoint xy;
124             dpiInfo(xy);
125             return tr("DPI: %1, %2")\&.arg(xy\&.x())\&.arg(xy\&.y());
126         }
127         default:
128             return tr("DPI stage %1")\&.arg(level);
129         }
130     } else if(prefix == "$light"){
131         switch(suffix){
132         case LIGHT_UP:
133         case LIGHT_UP_WRAP:
134             return "Brightness up";
135         case LIGHT_DOWN:
136         case LIGHT_DOWN_WRAP:
137             return "Brightness down";
138         }
139     } else if(prefix == "$lock"){
140         switch(suffix){
141         case LOCK_TOGGLE:
142             return "Toggle Windows lock";
143         case LOCK_ON:
144             return "Windows lock on";
145         case LOCK_OFF:
146             return "Windows lock off";
147         }
148     } else if(prefix == "$anim"){

src/ckb/keyaction.cpp
#include <qdebug.h>
#include "keyaction.h"
#include "kb.h"
#include "kbanim.h"
#include "kbprofile.h"
#include <QDateTime>
#include <QUrl>
#include <cstring>
#include <cstdio>
#ifdef USE_LIBX11
#include <X11/Xlib.h>
#endif

KeyAction::Type KeyAction::type() const {
    if(_value.isEmpty())
        return UNBOUND;
    if(_value.at(0) == '$')
        return SPECIAL;
    return NORMAL;
}

KeyAction::KeyAction(const QString &action, QObject *parent)
    : QObject(parent), _value(action), preProgram(0), relProgram(0), sniperValue(0)
{
}

KeyAction::KeyAction(QObject *parent)
    : QObject(parent), _value(""), preProgram(0), relProgram(0), sniperValue(0)
{
}

KeyAction::~KeyAction(){
    // Clean up processes
    if(preProgram){
        preProgram->kill();
        delete preProgram;
    }
    if(relProgram){
        relProgram->kill();
        delete relProgram;
    }
}

QString KeyAction::defaultAction(const QString& key){
    // G1-G18 are unbound by default
    if(key.length() >= 2 && key[0] == 'g'
		&& ((key.length() == 2 && key[1] >= '0' && key[1] <= '9')
		|| (key.length() == 3 && key[1] == '1' && key[2] >= '0' && key[2] <= '8')))
        return "";
    // So are thumbgrid buttons
    if(key.startsWith("thumb"))
        return "";
    // TODO: default action for MR
    if(key == "mr")
        return "";
    // M1-M3 switch modes
    if(key == "m1")
        return "$mode:0";
    if(key == "m2")
        return "$mode:1";
    if(key == "m3")
        return "$mode:2";
    // Brightness and Win Lock are their own functions
    if(key == "light")
        return "$light:2";
    if(key == "lock")
        return "$lock:0";
    // DPI buttons
    if(key == "dpiup")
        return "$dpi:-2";
    if(key == "dpidn")
        return "$dpi:-1";
    if(key == "sniper")
        return "$dpi:0";
    // Everything else is a standard keypress
    return key;
}

QString KeyAction::friendlyName(const KeyMap& map) const {
    if(_value.isEmpty())
        return "Unbound";
    QStringList parts = _value.split(":");
    QString prefix = parts[0];
    if(parts.length() < 2){
        KeyMap::Layout layout = map.layout();
        QString name = KeyMap::friendlyName(_value, layout);
        if(name.isEmpty())
            return "(Unknown)";
        return name;
    }
    int suffix = parts[1].toInt();
    if(prefix == "$mode"){
        switch(suffix){
        case MODE_PREV:
        case MODE_PREV_WRAP:
            return "Switch to previous mode";
        case MODE_NEXT:
        case MODE_NEXT_WRAP:
            return "Switch to next mode";
        default:
            return tr("Switch to mode %1").arg(suffix + 1);
        }
    } else if(prefix == "$dpi"){
        // Split off custom parameters (if any)
        int level = parts[1].split("+")[0].toInt();
        switch(level){
        case DPI_UP:
            return "DPI up";
        case DPI_DOWN:
            return "DPI down";
        case DPI_SNIPER:
            return "Sniper";
        case DPI_CUSTOM:{
            QPoint xy;
            dpiInfo(xy);
            return tr("DPI: %1, %2").arg(xy.x()).arg(xy.y());
        }
        default:
            return tr("DPI stage %1").arg(level);
        }
    } else if(prefix == "$light"){
        switch(suffix){
        case LIGHT_UP:
        case LIGHT_UP_WRAP:
            return "Brightness up";
        case LIGHT_DOWN:
        case LIGHT_DOWN_WRAP:
            return "Brightness down";
        }
    } else if(prefix == "$lock"){
        switch(suffix){
        case LOCK_TOGGLE:
            return "Toggle Windows lock";
        case LOCK_ON:
            return "Windows lock on";
        case LOCK_OFF:
            return "Windows lock off";
        }
    } else if(prefix == "$anim"){
        return "Start animation";
    } else if(prefix == "$program"){
        return "Launch program";
    } else if(prefix == "$macro"){
        return "Send G-key macro";
    }
    return "(Unknown)";
}

QString KeyAction::modeAction(int mode){
    return QString("$mode:%1").arg(mode);
}

QString KeyAction::dpiAction(int level, int customX, int customY){
    QString action = tr("$dpi:%1").arg(level);
    if(level == DPI_CUSTOM)
        action += tr("+%1+%2").arg(customX).arg(customY);
    return action;
}

QString KeyAction::lightAction(int type){
    return QString("$light:%1").arg(type);
}

QString KeyAction::lockAction(int type){
    return QString("$lock:%1").arg(type);
}

QString KeyAction::programAction(const QString& onPress, const QString& onRelease, int stop){
    // URL-encode the commands and place them in the string (":" and "+" are both replaced, so they won't interfere)
    return "$program:" + QString::fromUtf8(QUrl::toPercentEncoding(onPress.trimmed())) + "+" + QString::fromUtf8(QUrl::toPercentEncoding(onRelease.trimmed())) + QString("+%1").arg(stop);
}

const static int ANIM_ONCE = 0x1, ANIM_KRSTOP = 0x2;

QString KeyAction::animAction(const QUuid& guid, bool onlyOnce, bool stopOnRelease){
    int flags = (onlyOnce ? ANIM_ONCE : 0) | (stopOnRelease ? ANIM_KRSTOP : 0);
    return "$anim:" + guid.toString() + QString("+%1").arg(flags);
}

QString KeyAction::specialInfo(int& parameter) const {
    QStringList list = _value.split(":");
    if(list.length() < 2){
        parameter = INT_MIN;
        return "";
    }
    parameter = list[1].toInt();
    return list[0].replace("$", "");
}

int KeyAction::programInfo(QString& onPress, QString& onRelease) const {
    if(!isProgram())
        return 0;
    QString param = _value.mid(9);
    QStringList programs = param.split("+");
    if(programs.length() != 3)
        return 0;
    onPress = QUrl::fromPercentEncoding(programs[0].toUtf8());
    onRelease = QUrl::fromPercentEncoding(programs[1].toUtf8());
    return programs[2].toInt();
}

int KeyAction::dpiInfo(QPoint& custom) const {
    if(!isDPI())
        return 0;
    QString param = _value.mid(5);
    QStringList lxy = param.split("+");
    int level = lxy[0].toInt();
    if(level == DPI_CUSTOM){
        if(lxy.length() != 3)
            return 0;
        custom = QPoint(lxy[1].toInt(), lxy[2].toInt());
    }
    return level;
}

QUuid KeyAction::animInfo(bool &onlyOnce, bool &stopOnRelease) const {
    if(!isAnim())
        return QUuid();
    QString param = _value.mid(6);
    QStringList split = param.split("+");
    if(split.length() < 2)
        return QUuid();
    QUuid id = split[0];
    int flags = split[1].toInt();
    onlyOnce = !!(flags & ANIM_ONCE);
    stopOnRelease = !!(flags & ANIM_KRSTOP);
    return id;
}

QString KeyAction::driverName() const {
    if(isSpecial())
        return "";
    return _value;
}

void KeyAction::keyEvent(KbBind* bind, bool down){
    // No need to respond to standard actions
    if(!isSpecial())
        return;
    QStringList parts = _value.split(":");
    if(parts.length() < 2)
        return;
    QString prefix = parts[0];
    int suffix = parts[1].toInt();
    if(prefix == "$mode"){
        if(!down)
            return;
        // Change mode
        Kb* device = bind->devParent();
        KbProfile* currentProfile = device->currentProfile();
        int mode = currentProfile->indexOf(currentProfile->currentMode());
        int modeCount = currentProfile->modeCount();
        switch(suffix){
        case MODE_PREV_WRAP:
            mode--;
            if(mode < 0)
                mode = modeCount - 1;
            break;
        case MODE_NEXT_WRAP:
            mode++;
            if(mode >= modeCount)
                mode = 0;
            break;
        case MODE_PREV:
            mode--;
            break;
        case MODE_NEXT:
            mode++;
            break;
        default:
            // Absolute
            mode = suffix;
            break;
        }
        if(mode < 0 || mode >= modeCount)
            return;
        device->setCurrentMode(currentProfile->modes()[mode]);
    } else if(prefix == "$dpi"){
        KbPerf* perf = bind->perf();
        int level = parts[1].split("+")[0].toInt();
        switch(level){
        case DPI_UP:
            if(!down)
                return;
            perf->dpiUp();
            break;
        case DPI_DOWN:
            if(!down)
                return;
            perf->dpiDown();
            break;
        case DPI_SNIPER:
            if(down)
                sniperValue = perf->pushSniper();
            else {
                perf->popDpi(sniperValue);
                sniperValue = 0;
            }
            break;
        case DPI_CUSTOM:{
            QPoint xy;
            dpiInfo(xy);
            if(xy.x() <= 0 || xy.y() <= 0)
                break;
            if(down)
                sniperValue = perf->pushDpi(xy);
            else {
                perf->popDpi(sniperValue);
                sniperValue = 0;
            }
            break;
        }
        default:
            if(level < 1 || level >= KbPerf::DPI_COUNT
                    || !down)
                return;
            perf->dpi(level);
            break;
        }
    } else if(prefix == "$light"){
        if(!down)
            return;
        // Change brightness
        KbLight* light = bind->light();
        int dim = light->dimming();
        switch(suffix){
        case LIGHT_UP:
            if(dim > 0)
                dim--;
            break;
        case LIGHT_DOWN:
            if(dim < KbLight::MAX_DIM)
                dim++;
            break;
        case LIGHT_UP_WRAP:
            dim--;
            if(dim < 0)
                dim = KbLight::MAX_DIM;
            break;
        case LIGHT_DOWN_WRAP:
            dim++;
            if(dim > KbLight::MAX_DIM)
                dim = 0;
            break;
        }
        light->dimming(dim);
    } else if(prefix == "$lock"){
        if(!down)
            return;
        // Change win lock
        switch(suffix){
        case LOCK_TOGGLE:
            bind->winLock(!bind->winLock());
            break;
        case LOCK_ON:
            bind->winLock(true);
            break;
        case LOCK_OFF:
            bind->winLock(false);
            break;
        }
    } else if(prefix == "$anim"){
        // Start animation
        bool onlyOnce = false, stopOnRelease = false;
        QUuid id = animInfo(onlyOnce, stopOnRelease);
        KbAnim* anim = bind->light()->findAnim(id);
        if(!anim)
            return;
        if(down){
            if(!onlyOnce || !anim->isActive())
                // If "only once" is enabled, don't start the animation when it's already running
                anim->trigger(QDateTime::currentMSecsSinceEpoch(), true);
        } else if(stopOnRelease){
            // Key released - stop animation
            anim->stop();
        }
    } else if(prefix == "$program"){
        // Launch program
        QString onPress, onRelease;
        int stop = programInfo(onPress, onRelease);
        // Stop running programs based on setting
        QProcess* process = 0;
        if(down){
            if(stop & PROGRAM_PR_KPSTOP){
                process = preProgram;
                if(process)
                    process->kill();
                process = 0;
            }
            if(stop & PROGRAM_RE_KPSTOP)
                process = relProgram;
        } else {
            if(stop & PROGRAM_PR_KRSTOP)
                process = preProgram;
        }
        if(process)
            process->kill();
        // Launch new process if requested
        QString& program = down ? onPress : onRelease;
        if(program.isEmpty())
            return;
        // Check if the program is running already. If so, don't start it again.
        process = down ? preProgram : relProgram;
        if(process){
            if(process->state() == QProcess::NotRunning)
                delete process;
            else
                return;
        }

        // Adjust the selected display.
        adjustDisplay();

        // Start the program. Wrap it around sh to parse arguments.
        if((down && (stop & PROGRAM_PR_MULTI))
                || (!down && (stop & PROGRAM_RE_MULTI))){
            // Multiple instances allowed? Start detached process
            QProcess::startDetached("sh", QStringList() << "-c" << program);
        } else {
            process = new QProcess(this);
            process->start("sh", QStringList() << "-c" << program);
            if(down)
                preProgram = process;
            else
                relProgram = process;
        }
    } else if (prefix == "$macro") {
        // Do nothing, because all work is done by the keyboard itself.
        // For now, there is no reason to react on G-key press or release.
        // If u find some reason, then here is the place for it.
    }
}

/// \brief KeyAction::macroDisplay is just for debugging.
/// It shows the content of the key action and some other info.
///
void KeyAction::macroDisplay() {
    qDebug() << "isMacro returns" << (isMacro() ? "true" : "false");
    qDebug() << "isValidMacro returns" << (isValidMacro() ? "true" : "false");
    QStringList ret =_value.split(":");
    qDebug() << "Macro definition contains" << ret.count() << "elements";
    qDebug() << "Macro definition is" << _value;
}

void KeyAction::adjustDisplay(){
#ifdef USE_LIBX11
    // Try to get the current display from the X server
    char* display_name = XDisplayName(NULL);
    if(!display_name)
        return;
    Display* display = XOpenDisplay(display_name);
    if(!display)
        return;
    char* display_string = DisplayString(display);
    if(!display_string || strlen(display_string) == 0){
        XCloseDisplay(display);
        return;
    }
    size_t envstr_size = strlen(display_string) + 4;
    char* envstr = new char[envstr_size];
    strncpy(envstr, display_string, envstr_size);
    envstr[envstr_size - 1] = 0;

    Window root_window = XRootWindow(display, DefaultScreen(display));
    Window root_window_ret, child_window_ret, window;
    XWindowAttributes attr;
    int root_x, root_y, win_x, win_y;
    unsigned int mask_ret;

    // Find the screen which currently has the mouse
    XQueryPointer(display, root_window, &root_window_ret, &child_window_ret, &root_x, &root_y, &win_x, &win_y, &mask_ret);
    if(child_window_ret == (Window)NULL)
        window = root_window_ret;
    else
        window = child_window_ret;
    XGetWindowAttributes(display, window,  &attr);

    char* ptr = strchr(envstr, ':');
    if(ptr){
        ptr = strchr(ptr, '.');
        if(ptr)
            *ptr = '\0';
        char buf[16];
        snprintf(buf, sizeof(buf), ".%i", XScreenNumberOfScreen(attr.screen));
        strncat(envstr, buf, envstr_size - 1 - strlen(envstr));

        // Update environment variable
        setenv("DISPLAY", envstr, 1);
    }

    delete[] envstr;
    XCloseDisplay(display);
#endif
}

//////////
/// \brief KeyAction::macroAction is called when applying changes on a macro definition.
/// macroAction ist called while being in the macro pane
/// and clicking Apply with something in the Macro Text Box.
/// It tags that input with "$macro:" for further recognition.
/// \param macroDef holds the String containing parts 2-5 of a complete macro definition.
/// \return QString holding the complete G-Key macro definition (parts 1-5)
///
QString KeyAction::macroAction(QString macroDef) {
    return QString ("$macro:%1").arg(macroDef);
}
